File: csharp/1475.final-prices-with-a-special-discount-in-a-shop.cs
/*
 * @lc app=leetcode id=1475 lang=csharp
 *
 * [1475] Final Prices With a Special Discount in a Shop
 */

using System.Collections.Generic;

// @lc code=start
public class Solution {
    public int[] FinalPrices(int[] prices) {
        int n = prices.Length;
        int[] ret = new int[n];
        var stack = new Stack<int>();
        for (int i = 0; i < n; ++i) {
            int price = prices[i];
            ret[i] = price;
            while (stack.TryPeek(out var j) && prices[j] >= price) {
                ret[j] -= price;
                stack.Pop();
            }
            stack.Push(i);
        }
        return ret;
    }
}
// @lc code=end

File: java/1475.final-prices-with-a-special-discount-in-a-shop.java
/*
 * @lc app=leetcode id=1475 lang=java
 *
 * [1475] Final Prices With a Special Discount in a Shop
 */

import java.util.Stack;

// @lc code=start
class Solution {
    public int[] finalPrices(int[] prices) {
        Stack<Integer> stack = new Stack<>();
        int n = prices.length;
        stack.push(0);
        for (int i = n - 1; i >= 0; --i) {
            int p = prices[i];
            while (p < stack.peek()) {
                stack.pop();
            }
            prices[i] -= stack.peek();
            stack.push(p);
        }
        return prices;
    }
}
// @lc code=end

